python/src/tasks/tasks.py

```diff
@@ -41,31 +53,36 @@ def submit(self, msg, adata):
                         error=True,
                     )
                 ]
-            return result, adata
+            return result
+
+    def _factory(self, msg):
+        print("before checking for change: ", self.adata_path)
+        if count_matrix.is_file_changed(self.adata_path):
+            print("There has been recent write to the file, have to download it again.")
+            # TODO: after introducing multiple-sample support, make this
+            #  more efficient by making it able to download only a specific file too.
+            count_matrix.download_files()
+            self._initialise_adata()
```

Comment:
I think this is still too difficult to read, but I think now reading all the code I have a better understanding of what you were doing. Here is a possibly better way to go, and this lends itself nicely to the R development that is going to be carried out immediately.

Expose just one function, `download_files()` from `count_matrix`. It does the following:

1. check if the path `/data/$EXPERIMENT_ID` exists, save it to a variable called `path_exists`.
2. call boto with `list_objects`, save it to a variable called `objects`.
3. for each object in `objects`, call `_calculate_file_etag()`. write it so it handles files not existing, and in that case return `None` if return value of `_calculate_file_etag()` does not match, call `download_file(object_name)`, which will download the file from the object path `213213asdas2134213/python.h5ad` into `/data/213213asdas2134213/python.h5ad`, as appropriate.

we can jump on a call and I can show you what i mean to discuss this.